### PR TITLE
[2pt] Add a table with feature comparison of Python connectors

### PR DESCRIPTION
**Pull request by ligurio**
Patch adds a table with comparison of Python connectors for Tarantool:
- asynctnt
- gtarantool
- tarantool-python

Dmitry Shveenkov (@shveenkov), Igor Latkin (@igorcoding), FYI.

### CODE REVIEW

**Comment by Totktonada** [comment]
image

https://develop.tarantool.io/en/doc/python-connectors/book/connectors/#feature-comparison

@NickVolynkin Can we do something with that?

**Comment by Totktonada** [comment]
(And links in the last column are no clickable.)

File: doc/book/connectors/__python.rst
Comment (Totktonada) [inline doc/book/connectors/__python.rst:66]: I guess the Simplified BSD license (BSD-2-Clause) was meant and the Modified BSD license (BSD-3-Clause) was referred by a mistake. I would like to resolve this ambiguity: https://github.com/tarantool/tarantool-python/issues/197.

File: doc/book/connectors/__python.rst
Comment (Totktonada) [inline doc/book/connectors/__python.rst:98]: It would be fair to say 'Travis CI (disabled) and Appveyor (only Windows)'.

File: doc/book/connectors/__python.rst
Comment (Totktonada) [inline doc/book/connectors/__python.rst:130]: I think we can say 'Yes, gevent' without the question mark.

File: doc/book/connectors/__python.rst
Comment (Totktonada) [inline doc/book/connectors/__python.rst:167]: AFAIR, tarantool-python has some support of varbinary (but there are questions how it is mapped into python types). I would also re-check other connectors here.

(It is more a note for myself.)

File: doc/book/connectors/__python.rst
Comment (Totktonada) [inline doc/book/connectors/__python.rst:245]: AFAIR, the `cluster_discovery_function` argument is about renewing the list of URI for the connection pool and does not provide ability to differentiate a master from a replica. There is https://github.com/tarantool/tarantool-python/issues/196 about master discovering.

File: doc/book/connectors/__python.rst
Comment (Totktonada) [inline doc/book/connectors/__python.rst:257]: We can refer https://github.com/tarantool/tarantool-python/wiki/PEP-249-Database-API — I think it works good as the introduction, but it also highlight details a user may need to know about the driver.

File: doc/book/connectors/__python.rst
Comment (Totktonada) [inline doc/book/connectors/__python.rst:251]: In this wording it looks confusing: MeshConnection has a connection pool under the hood.

**Comment by Totktonada** [comment]
I have several more things I would like to clarify, but I want to dive a bit into the context and come back with more details.

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst:98]: I'm not sure, it is a meaningful detail for Tarantool users.

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst:257]: Yes, good idea.

File: doc/book/connectors/__python.rst
Comment (DifferentialOrange) [inline doc/book/connectors/__python.rst:54]: You can also add `:stub-columns: 1` to change style of the left column

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst:257]: Added a link, thanks!

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst:130]: removed question mark

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst:66]: Okay, I'll change it as soon as PR will be merged.

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst:54]: Added

**Comment by ligurio** [comment]
Rename "Estimation of importance of known bugs" -> "Known Issues".

File: doc/book/connectors/__python.rst
Comment (Totktonada) [reply doc/book/connectors/__python.rst:100]: It allows to make a guess about code healthy and popularity of the project, so looks useful. At least if we'll fairly deliver the actual state, which is 'broken Travis CI and Appveyor for Windows' now.

File: doc/book/connectors/__python.rst
Comment (Totktonada) [inline doc/book/connectors/__python.rst:143]: There is no such exception in asynctnt. What is the idea behind this note?

File: doc/book/connectors/__python.rst
Comment (Totktonada) [reply doc/book/connectors/__python.rst:169]: We can say 'questionable' and link https://github.com/tarantool/tarantool-python/issues/105.

File: doc/book/connectors/__python.rst
Comment (Totktonada) [inline doc/book/connectors/__python.rst:189]: I would get rid of 'Soon' marks on this page.

File: doc/book/connectors/__python.rst
Comment (Totktonada) [inline doc/book/connectors/__python.rst:243]: I would not link this issue here, see https://github.com/tarantool/doc/pull/2484#discussion_r772349714.

File: doc/book/connectors/__python.rst
Comment (Totktonada) [inline doc/book/connectors/__python.rst:156]: Is there a difference between 'No (using call server-side functions)' and 'No'?

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst:68]: Changed to BSD-2 as discussed.

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst:169]: Discussed and decided to recheck varbinary support in tarantool-python.

File: doc/book/connectors/__python.rst
Comment (ligurio) [inline doc/book/connectors/__python.rst:60]: Decided to remove aiotarantool from the table.

File: doc/book/connectors/__python.rst
Comment (Totktonada) [reply doc/book/connectors/__python.rst:143]: asynctnt is the first one, my mistake.

File: doc/book/connectors/__python.rst
Comment (Totktonada) [inline doc/book/connectors/__python.rst:235]: Let's remove question marks.

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst:189]: Without mark, it looks as existed Tarantool's feature. Okey, removed "Soon" marks.

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst:243]: Could you clarify what do you mean by that link (GH failed to show a comment body in PR if comment was resolved)?

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst]: ok

File: doc/book/connectors/__python.rst
Comment (Totktonada) [reply doc/book/connectors/__python.rst]: image

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst]: Replaced with 'Yes'.

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst]: The short explanation for those who will read it: there is a 'master discovery' used in connectors, and there is another 'master discovery' that PM's planned to implement in Tarantool. To avoid confusion, we don't need to mix up both 'master discovery'. It is a reason why link should be removed.

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst]: Updated

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst]: Fixed

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst]: In another [comment](https://github.com/tarantool/doc/pull/2482#discussion_r776313461), we decided to don't mix 'master discovery' in connectors and 'master discovery' in Tarantool.

Comment (Totktonada) [reply doc/book/connectors/__python.rst]: Yep, but this comment still relevant: we have no master discovery in tarantool-python, but the table states the opposite.

Comment (Totktonada) [reply doc/book/connectors/__python.rst]: (And, while we're here, 'connection pool' could be removed from the parentheses, because we already say 'Yes' for the 'Connection pool' row.)

File: doc/book/connectors/__python.rst
Comment (Totktonada) [inline doc/book/connectors/__python.rst:139]: A simple test does not show it as supported for tarantool-python:

```lua
tarantool> box.cfg{listen = 3301}
tarantool> unpack(box.space.s:format())
---
- {'type': 'unsigned', 'name': 'id', 'version': 2}
- {'type': 'string', 'name': 'd', 'is_nullable': true}
- {'type': 'map', 'name': 'x', 'is_nullable': true}
...
```

```python
>>> import tarantool
>>> c = tarantool.connect('localhost', 3301)
>>> ts = c.select('s')
>>> t = ts[1]
>>> t
[2]
>>> t.id
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'list' object has no attribute 'id'
>>> t['id']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: list indices must be integers or slices, not str
```

AFAIR asynctnt has a tuple object and so likely supports it (but I didn't check).

File: doc/book/connectors/__python.rst
Comment (Totktonada) [inline doc/book/connectors/__python.rst:50]: I would leave a reason why aiotarantool is not in the table.

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:42]:
```suggestion
Also, there are several community-driven Python connectors (see the table below with
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:76]:
```suggestion
        -   `issue #105 <https://github.com/tarantool/tarantool-python/issues/105>`__ (unpack of binary data)
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:89]:
```suggestion
        -   GitHub Actions
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:90]:
```suggestion
        -   No (tests exist)
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:91]:
```suggestion
        -   AppVeyor (Windows only)
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:93]:
```suggestion
    *   -   GitHub Stars
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:47]:
```suggestion
* `gtarantool <https://github.com/shveenkov/gtarantool>`__ with gevent support, **no active maintenance**
```

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst:50]: Done.

```diff
+++ b/doc/book/connectors/__python.rst
@@ -39,17 +39,20 @@ For an example of using Python API with
 `queue managers for Tarantool <https://github.com/tarantool/queue>`__, see
 `queue-python <https://github.com/tarantool/queue-python>`__ project at GitHub.
 
-Also there are several community-driven Python connectors (see table below with
-feature comparison):
+Also there are several community-driven Python connectors:
 
 * `asynctnt <https://github.com/igorcoding/asynctnt>`__ with asyncio support
 * `aiotarantool <https://github.com/shveenkov/aiotarantool>`__ also with asyncio support
 * `gtarantool <https://github.com/shveenkov/gtarantool>`__ with gevent support **no active maintenance**
 
+Table below contains a feature comparison for asynctnt, gtarantool and
+tarantool-python. aiotarantool is absent there because it is quite outdated and
+unmaintained.
+
 Feature comparison
 ------------------
```

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst:139]: You are right! Thanks, fixed.
```diff
@@ -134,7 +137,7 @@ Last update: January 2022
         -   Yes
 
     *   -   Access tuple fields by names
-        -   Yes
+        -   No
         -   Yes
         -   Yes
 
```

Comment (Totktonada) [reply doc/book/connectors/__python.rst]: gtarantool is based on tarantool-python and likely the same?

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:46]:
```suggestion
* `gtarantool <https://github.com/shveenkov/gtarantool>`__ with gevent support, **no active maintenance**
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:48]:
```suggestion
The table below contains a feature comparison for asynctnt, gtarantool and
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:79]:
```suggestion
        -   `issue #105 <https://github.com/tarantool/tarantool-python/issues/105>`__ (unpack of binary data)
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:92]:
```suggestion
        -   GitHub Actions
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:93]:
```suggestion
        -   No (tests exist)
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:94]:
```suggestion
        -   AppVeyor (only Windows)
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:96]:
```suggestion
    *   -   GitHub Stars
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:210]:
```suggestion
        -   Autoreconnect
```

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst]: Removed 'connection pool' and 'master discovery' from parentheses.

File: doc/book/connectors/__python.rst
Comment (ligurio) [reply doc/book/connectors/__python.rst]: Yes, set for it 'No' too.

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:46]:
```suggestion
* `gtarantool <https://github.com/shveenkov/gtarantool>`__ with gevent support, **no active maintenance**
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:48]:
```suggestion
The table below contains a feature comparison for asynctnt, gtarantool and
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:79]:
```suggestion
        -   `issue #105 <https://github.com/tarantool/tarantool-python/issues/105>`__ (unpack of binary data)
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:92]:
```suggestion
        -   GitHub Actions
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:93]:
```suggestion
        -   No (tests exist)
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:94]:
```suggestion
        -   AppVeyor (only Windows)
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:96]:
```suggestion
    *   -   GitHub Stars
```

File: doc/book/connectors/__python.rst
Comment (arctic-dreamer) [inline doc/book/connectors/__python.rst:210]:
```suggestion
        -   Autoreconnect
```

**Review by Totktonada — APPROVED** [review]
LGTM from my side. Please, proceed with the documentation team.

**Comment by Totktonada** [comment]
Please, ask me for review after the translation will be ready.

File: doc/book/connectors/__python.rst
Comment (Totktonada) [inline doc/book/connectors/__python.rst:52]: I don't know why we have two docs regarding connectors in different sections (that's odd and confusing), but as result we end up with the comparison tables in different sections (see PR #2484).

File: locale/ru/LC_MESSAGES/book/connectors.po
Comment (Totktonada) [inline locale/ru/LC_MESSAGES/book/connectors.po:1004]: Здесь стоит выровнять с тем, как будет в PR #2484 (см. мои комментарии).

**Review by Totktonada — APPROVED** [review]
LGTM after resolving comments above (no need to re-review with me after the changes).

File: doc/book/connectors/__python.rst
Comment (patiencedaur) [reply doc/book/connectors/__python.rst:52]: We noticed that and have filed an issue: https://github.com/tarantool/doc/issues/2591

**Comment by ligurio** [comment]
Probably we should add a link to comparisons of Go/Python connectors on page https://www.tarantool.io/ru/download/connectors. @Totktonada, @DifferentialOrange, what do you think?

**Comment by DifferentialOrange** [comment]
> Probably we should add a link to comparisons of Go/Python connectors on page https://www.tarantool.io/ru/download/connectors. @Totktonada, @DifferentialOrange, what do you think?

I think we have too many different connector pages and it wasn't planned to be like this

**Comment by Totktonada** [comment]
+1 to Georgy. While it is in such unclear state, I can't find any logic here and so can't think around.

@patiencedaur Is the problem on track?

**Comment by patiencedaur** [comment]
Just to be clear, too many different connector pages is two pages?
https://www.tarantool.io/ru/download/connectors
https://www.tarantool.io/ru/doc/latest/book/connectors/

Please let me know if there are any more connector-related pages you'd like to see squashed. After that, I'll create an issue.

**Comment by Totktonada** [comment]
Those two are the same. But there are two different ones. Say, for tarantool-python:

```
doc/getting_started/getting_started_python.rst
doc/book/connectors/__python.rst
```

Or for go-tarantool:

```
doc/getting_started/getting_started_go.rst
doc/book/connectors/__go.rst
```

And likely others.

**Comment by patiencedaur** [comment]
Thank you! It will most likely be part of our restructuring epic, which is now planned for Q3–Q4.

**Pull request merged**
